v1/plugins/logs/eventBuffer.go

```diff
@@ -0,0 +1,240 @@
+		}
+		push(newBuffer, event)
+	}
+	b.buffer = newBuffer
+}
+
+// Push attempts to add a new event to the buffer
+func (b *eventBuffer) Push(event EventV1) error {
+	encoded, err := json.Marshal(event)
+	if err != nil {
+		return err
+	}
+
+	err = b.dropNDCache(&event, encoded)
+	if err != nil {
+		if !errors.As(err, &droppedNDCache{}) {
+			return err
+		}
+	}
+
+	// This only blocks when the buffer is being reconfigured
+	b.mtx.RLock()
+	defer b.mtx.RUnlock()
+
+	push(b.buffer, encoded)
+
+	return err
+}
+
+// dropNDCache checks if an event is bigger than UploadSizeLimitBytes and
+// tries to drop the NDBuiltinCache. If the event is still too big, drop the event.
+func (b *eventBuffer) dropNDCache(event *EventV1, encoded []byte) error {
+	if int64(len(encoded)) < b.uploadSizeLimitBytes {
+		return nil
+	}
+
+	if event.NDBuiltinCache == nil {
+		return fmt.Errorf("upload chunk size (%d) exceeds upload_size_limit_bytes (%d)",
+			int64(len(encoded)), b.uploadSizeLimitBytes)
+	}
+	event.NDBuiltinCache = nil
+	encoded, err := json.Marshal(event)
+	if err != nil {
+		return err
+	}
+	if int64(len(encoded)) > b.uploadSizeLimitBytes {
+		return fmt.Errorf("upload chunk size (%d) exceeds upload_size_limit_bytes (%d)",
+			int64(len(encoded)), b.uploadSizeLimitBytes)
+	}
+
+	return droppedNDCache{}
+}
+
+func push[T any](ch chan T, data T) {
+	select {
+	case ch <- data:
+	default:
+		<-ch
+		ch <- data
+	}
+}
+
+// Upload reads all the currently buffered events and uploads them as a gzip compressed JSON array to the service.
+// The events are uploaded in chunks limited by the uploadSizeLimitBytes.
+func (b *eventBuffer) Upload(ctx context.Context) (bool, error) {
+	b.payload.mtx.Lock()
+	defer b.payload.mtx.Unlock()
+
+	// This only blocks when the buffer is being reconfigured
+	b.mtx.RLock()
+	defer b.mtx.RUnlock()
+
+	events := len(b.buffer)
+	for range events {
+		select {
+		case <-ctx.Done():
+			return false, ctx.Err()
+		case event := <-b.buffer:
+			if !b.payload.capacity(len(event), b.uploadSizeLimitBytes) {
+				ok, err := b.payload.upload(ctx, b.client, b.uploadPath)
+				if !ok || err != nil {
+					return false, err
+				}
+			}
+
+			if err := b.payload.add(event); err != nil {
+				return false, err
+			}
+		default:
+			break
+		}
+	}
+
+	return b.payload.upload(ctx, b.client, b.uploadPath)
+}
+
+// payloadBuffer contains the compressed JSON array to be uploaded
+type payloadBuffer struct {
+	buffer       *bytes.Buffer
+	mtx          sync.Mutex
+	writer       *gzip.Writer
+	closed       bool
+	bytesWritten int
+}
+
+func newPayloadBuffer() *payloadBuffer {
+	p := payloadBuffer{}
+	p.buffer = new(bytes.Buffer)
+	p.writer = gzip.NewWriter(p.buffer)
+	return &p
+}
+
+func (p *payloadBuffer) capacity(eventLen int, uploadSizeLimitBytes int64) bool {
+	return !(int64(eventLen+p.bytesWritten+1) > uploadSizeLimitBytes)
+}
+
+// add writes an event to the JSON array
+func (p *payloadBuffer) add(event []byte) error {
+	switch p.bytesWritten {
```

Comment:
Just for clarification -- there's only a single process ever calling p.add, so we don't need to consider concurrency here? There can't be two "first events" hitting payloadBuffer concurrently?

Reply:
Correct, `add` is only called in one place in the `Upload` function which uses the mutex `b.payload.mtx.Lock()` to control uploads happen sequentially.

Reply:
Thanks!